app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,197 @@
+  # see https://github.com/3scale/activejob-uniqueness/blob/main/lib/active_job/uniqueness/strategies/until_executed.rb
+  unique :until_and_while_executing, lock_ttl: 6.hours
 
-  def on_complete(_, options)
-    on_finish('on_complete', options)
-  end
+  # better limit by available  `delete` executors
+  # sidekiq_throttle concurrency: { limit: 10 }
 
-  def on_finish(method_name, options)
-    workers_hierarchy = options['caller_worker_hierarchy']
-    info "Starting DeleteObjectHierarchyWorker##{method_name} with the hierarchy of workers: #{workers_hierarchy}"
-    object = GlobalID::Locator.locate(options['object_global_id'])
-    background_destroy_method = @background_destroy_method.presence || 'destroy'
-    DeletePlainObjectWorker.perform_later(object, workers_hierarchy, background_destroy_method)
-    info "Finished DeleteObjectHierarchyWorker##{method_name} with the hierarchy of workers: #{workers_hierarchy}"
-  rescue ActiveRecord::RecordNotFound => exception
-    info "DeleteObjectHierarchyWorker##{method_name} raised #{exception.class} with message #{exception.message}"
+  before_perform do |job|
+    info "Starting #{job.class}#perform with the hierarchy of workers: #{job.arguments}"
   end
 
-  protected
-
-  delegate :info, to: 'Rails.logger'
+  after_perform do |job|
+    if @remaining_hierarchy.present?
+      self.class.perform_later(*@remaining_hierarchy)
+      msg = " iteration of"
+    end
+    info "Finished#{msg} #{job.class}#perform with the hierarchy of workers: #{job.arguments}"
+  end
 
-  attr_reader :object, :caller_worker_hierarchy
+  class << self
+    # convenience method to schedule deleting an active record object
+    def delete_later(ar_object)
+      perform_later *hierarchy_entries_for(ar_object)
+    end
 
-  def build_batch
-    batch = Sidekiq::Batch.new
-    batch.description = batch_description
-    batch_callbacks(batch) { batch.jobs { destroy_and_delete_associations } }
-    batch
-  end
+    private
 
-  def batch_description
-    "Deleting #{object.class.name} [##{object.id}]"
-  end
+    # @return the hierarchy entries to handle deletion of an object
+    def hierarchy_entries_for(ar_object)
+      return [] unless ar_object&.persisted? # e.g. when calling Proxy#oidc_configuration a new object can be generated
+
+      if ar_object.respond_to?(:destroyable?, true) && !ar_object.send(:destroyable?)
+        raise DoNotRetryError, "Background deleting #{ar_object.class}:#{ar_object.id} which is not destroyable."
+      elsif ar_object.is_a?(FeaturesPlan)
+        # This is an ugly hack to handle lack of `#id` but we have only FeaturesPlans with a composite primary key.
+        # Rails 7.1 supports composite primary keys so we can implement universal handling. See [FeaturesPlan].
+        # Now to avoid complications, just sweep it under the rag.
+        FeaturesPlan.where(feature_id: ar_object.feature_id, plan_id: ar_object.plan_id).delete_all
```

Comment:
So, this is not the same as `ar_object.delete`?

Reply:
I can assure you, I wouldn't write this crap if Rails 7.0 supported composite keys and `ar_object.delete` worked. I'm actually not sure whehter composite keys were added in 7.1 or 7.3. Anyway, this needs to stay like this for the time being.

P.S. fun fact that adding `limit(1)` to the statement works on mysql but breaks postgres. So there is no limit to the statement. But that should be totally fine.